src/shims/collectionShims.ts

```diff
@@ -0,0 +1,443 @@
+        keys(): IteratorShim<T>;
+        values(): IteratorShim<T>;
+        entries(): IteratorShim<[T, T]>;
+        forEach(action: (value: T, key: T) => void): void;
+    }
+
+    interface SetShim<T> extends ReadonlySetShim<T> {
+        add(value: T): this;
+        delete(value: T): boolean;
+        clear(): void;
+    }
+
+    type SetShimConstructor = new <T>(iterable?: readonly T[] | ReadonlySetShim<T>) => SetShim<T>;
+
+    interface MapData<K, V> {
+        size: number;
+        strings?: Record<string, MapEntry<K, V>>;
+        numbers?: MapEntry<K, V>[];
+        nan?: MapEntry<K, V>;
+        positiveInfinity?: MapEntry<K, V>;
+        negativeInfinity?: MapEntry<K, V>;
+        list?: MapEntry<K, V>;
+        true?: MapEntry<K, V>;
+        false?: MapEntry<K, V>;
+        null?: MapEntry<K, V>;
+        undefined?: MapEntry<K, V>;
+
+        // Linked list references for iterators.
+        // See https://github.com/Microsoft/TypeScript/pull/27292
+        // for more information.
+
+        /**
+         * The first entry in the linked list.
+         * Note that this is only a stub that serves as starting point
+         * for iterators and doesn't contain a key and a value.
+         */
+        readonly first: MapEntry<K, V>;
+        last: MapEntry<K, V>;
+    }
+
+    interface MapEntry<K, V> {
+        readonly key?: K;
+        value?: V;
+
+        // Linked list references for non-primitive keys
+        next?: MapEntry<K, V>;
+
+        // Linked list references for iterators.
+        nextIterationEntry?: MapEntry<K, V>;
+        previousIterationEntry?: MapEntry<K, V>;
+
+        /**
+         * Specifies if iterators should skip the next entry.
+         * This will be set when an entry is deleted.
+         * See https://github.com/Microsoft/TypeScript/pull/27292 for more information.
+         */
+        skipNextIteration?: boolean;
+    }
+
+    interface IteratorData<K, V, U extends (K | V | [K, V])> {
+        currentEntry?: MapEntry<K, V>;
+        selector: (key: K, value: V) => U;
+    }
+
+    function createDictionaryObject<T>(): Record<string, T> {
+        const map = Object.create(/*prototype*/ null); // eslint-disable-line no-null/no-null
+
+        // Using 'delete' on an object causes V8 to put the object in dictionary mode.
+        // This disables creation of hidden classes, which are expensive when an object is
+        // constantly changing shape.
+        map.__ = undefined;
+        delete map.__;
+        return map;
+    }
+
+    function createMapData<K, V>(): MapData<K, V> {
+        const entry: MapEntry<K, V> = {};
+        return { first: entry, last: entry, size: 0 };
+    }
+
+    function createMapEntry<K, V>(key: K, value: V): MapEntry<K, V> {
+        return { key, value, nextIterationEntry: undefined, previousIterationEntry: undefined, skipNextIteration: false };
+    }
+
+    function createListMapEntry<K, V>(key: K, value: V, next?: MapEntry<K, V>): MapEntry<K, V> {

```

Comment:
While allowing different key types seems reasonable, we should do that when we really need it.. We should currently implement only for `T extends string ` keys to keep this simple and compact

Reply:
Infact thats the missing piece.. Previously we ensured we are using string as keys and now we are missing that ability to ensure and make conscious choice of whether we really need map that doesnt key on string

Reply:
One of the goals of this PR is specifically to allow us to use non-string keys for both `Map` and `Set`, as we have a number of cases where we coerce numbers like symbol IDs, type IDs, and node IDs to string, which is unnecessary overhead. In some cases, we will even be able to use `Map<Node, ...>` or `Set<Node>` instead of allocating a new node ID, and remove the extra cost of those calculations as well.

Reply:
As I've mentioned elsewhere, we already have a `NodeMap` and `NodeSet` class in _utilities.ts_ that has to work around `Map` only supporting `string` keys to act as if it were a native map that allows `Node` as a key. Since we're using a native `Map` under the covers, the custom `NodeMap` and `NodeSet` classes really aren't very useful. #39264 builds on this PR and removes `NodeMap` and `NodeSet` in favor of a regular `Map` and `Set`.